### DISCUSSION TITLE
[@types/three] Unify typeof .userData property

### DISCUSSION THREAD

**Discussion by hanvit-eazel** [DefinitelyTyped/DefinitelyTyped#67213] [Issues with a @types package] [answered]
### Description

Issue submitted in three.js first
[link](https://github.com/mrdoob/three.js/issues/27070)

Here are the classes that currently use the userData property in three.js.
(captured from @types/three r155.1)

<img width="520" alt="hasUserData" src="https://github.com/mrdoob/three.js/assets/82128390/0bb2f1c6-a338-4e2c-bbe2-7b0b38957faf">

- Object3D
- BufferGeometry
- Material
- Texture
- (UserDataNode, GLTF)

But as you can see, Object3D, BufferGeometry and Material, Texture have different userData types.

This situation caused me the following type issue.

<img width="570" alt="error" src="https://github.com/mrdoob/three.js/assets/82128390/2e629873-132b-45aa-870e-94867dac5910">

```
type UserDataWrappedObject3D = Object3D & {
    userData: { myCustomValue: number }
}

type UserDataWrappedMaterial = Material & {
    userData: { myCustomValue: number }
}

let a: UserDataWrappedObject3D 
let b: UserDataWrappedMaterial 

a.userData.myCustomValue
b.userData.myCustomValue // error from userData inferred as any
```

### Solution

If we want to allow wrapping of userData in the above manner, I suggest replacing the userData type of Materials and Textures with Object3D or BufferGeometry's userData type.

```
// in Material class
...
userData: { [key: string]: any };
...
```

### Alternatives

Or we can apply `Record` type for userData as it commented in code.

[a comment by an automated account is not shown]

**Comment by Methuselah96** [comment, marked as answer]
Hey @hanvit-eazel, thanks for the issue! Do you mind reposing it as an issue on https://github.com/three-types/three-ts-types, that will make it easier to keep track of. Thanks!

**Reply by hanvit-eazel** [reply]
Thanks!

[issue in three-types](https://github.com/three-types/three-ts-types/issues/656)
